<a href="https://colab.research.google.com/github/lililalina/UoM-stats-with-python/blob/main/02%20Inferential%20Statistics/3_5_week_3_assesment_hypothesis_testing_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Hypothesis Tests in Python
In this assessment, you will look at data from a study on toddler sleep habits. 

The hypothesis tests you create and the questions you answer in this Jupyter notebook will be used to answer questions in the following graded assignment.

In [76]:
import numpy as np
import pandas as pd
from scipy.stats import t
pd.set_option('display.max_columns', 30) # set so can see all columns of the DataFrame

Your goal is to analyse data which is the result of a study that examined differences in a number of sleep variables between napping and non-napping toddlers. Some of these sleep variables included: Bedtime (lights-off time in decimalized time), Night Sleep Onset Time (in decimalized time), Wake Time (sleep end time in decimalized time), Night Sleep Duration (interval between sleep onset and sleep end in minutes), and Total 24-Hour Sleep Duration (in minutes). Note: Decimalized time is the representation of the time of day using units which are decimally related.

The 20 study participants were healthy, normally developing toddlers with no sleep or behavioral problems. These children were categorized as napping or non-napping based upon parental report of children’s habitual sleep patterns. Researchers then verified napping status with data from actigraphy (a non-invasive method of monitoring human rest/activity cycles by wearing of a sensor on the wrist) and sleep diaries during the 5 days before the study assessments were made.

You are specifically interested in the results for the Bedtime and Total 24-Hour Sleep Duration.

Reference: Akacem LD, Simpkin CT, Carskadon MA, Wright KP Jr, Jenni OG, Achermann P, et al. (2015) The Timing of the Circadian Clock and Sleep Differ between Napping and Non-Napping Toddlers. PLoS ONE 10(4): e0125181. https://doi.org/10.1371/journal.pone.0125181

In [77]:
# Import the data
df = pd.read_csv("/content/drive/MyDrive/COLAB NOTEBOOK/UoM stats with python/napping.csv") 

In [78]:
# First, look at the DataFrame to get a sense of the data
df

id     sex  age (months)  dlmo time  \
0                              1  female          33.7      19.24   
1                              2  female          31.5      18.27   
2                              3    male          31.9      19.14   
3                              4  female          31.6      19.69   
4                              5  female          33.0      19.52   
5                              6  female          36.2      18.22   
6                              7    male          36.3      19.28   
7                              8    male          30.0      21.06   
8                              9    male          33.2      19.38   
9                             10  female          37.1      19.93   
10                            11    male          32.9      18.79   
11                            12  female          35.0      19.65   
12                            13    male          35.1      19.83   
13                            14  female          35.6      19.88   
14                            15  female          36.6      19.94   
15                            16    male          36.5      20.25   
16                            17  female          33.7      20.33   
17                            18    male          36.4      20.16   
18                            19  female          33.6      19.68   
19                            20    male          33.8      20.51   
20                           NaN     NaN           NaN        NaN   
21               0 = no; 1 = yes     NaN           NaN        NaN   
22  all times are decimal values     NaN           NaN        NaN   

    days napped  napping  nap lights outl time  nap sleep onset  nap midsleep  \
0           0.0      0.0                   NaN              NaN           NaN   
1           0.0      0.0                   NaN              NaN           NaN   
2           0.0      0.0                   NaN              NaN           NaN   
3           0.0      0.0                   NaN              NaN           NaN   
4           0.0      0.0                   NaN              NaN           NaN   
5           4.0      1.0                 14.00            14.22         15.00   
6           1.0      1.0                 14.75            15.03         15.92   
7           5.0      1.0                 13.09            13.43         14.44   
8           2.0      1.0                 14.41            14.42         15.71   
9           3.0      1.0                 13.12            13.42         14.31   
10          4.0      1.0                 13.99            14.03         14.85   
11          5.0      1.0                 13.18            13.45         14.33   
12          3.0      1.0                 13.94            14.48         15.26   
13          4.0      1.0                 12.68            13.08         13.92   
14          4.0      1.0                 12.71            12.88         13.80   
15          3.0      1.0                 13.74            14.68         15.66   
16          5.0      1.0                 13.15            13.87         14.49   
17          5.0      1.0                 12.47            12.56         13.30   
18          3.0      1.0                 14.71            14.85         15.46   
19          3.0      1.0                 12.68            13.54         14.30   
20          NaN      NaN                   NaN              NaN           NaN   
21          NaN      NaN                   NaN              NaN           NaN   
22          NaN      NaN                   NaN              NaN           NaN   

    nap sleep offset  nap wake time  nap duration  nap time in bed  \
0                NaN            NaN           NaN              NaN   
1                NaN            NaN           NaN              NaN   
2                NaN            NaN           NaN              NaN   
3                NaN            NaN           NaN              NaN   
4                NaN            NaN           NaN              NaN   
5       

**Question**: What value is used in the column 'napping' to indicate a toddler takes a nap? (see reference article) 
**Answer** : napping = 1, no napping = 0


**Questions**: What is the overall sample size $n$? What are the sample sizes of napping and non-napping toddlers?
**Answer** : napping = 15, non-napping = 5

## Hypothesis tests
We will look at two hypothesis test, each with $\alpha = .05$:  


1. Is the average bedtime for toddlers who nap later than the average bedtime for toddlers who don't nap?


$$H_0: \mu_{nap}=\mu_{no\ nap}, \ H_a:\mu_{nap}>\mu_{no\ nap}$$
Or equivalently:
$$H_0: \mu_{nap}-\mu_{no\ nap}=0, \ H_a:\mu_{nap}-\mu_{no\ nap}>0$$


2. The average 24 h sleep duration (in minutes) for napping toddlers is different from toddlers who don't nap.


$$H_0: \mu_{nap}=\mu_{no\ nap}, \ H_a:\mu_{nap}\neq\mu_{no\ nap}$$
Or equivalently:
$$H_0: \mu_{nap}-\mu_{no\ nap}=0, \ H_a:\mu_{nap}-\mu_{no\ nap} \neq 0$$

First isolate `night bedtime` into two variables - one for toddlers who nap and one for toddlers who do not nap.

In [79]:
nap_bedtime = df['night bedtime'][df['napping'] == 1]
nap_bedtime

5     19.95
6     20.60
7     22.01
8     20.24
9     20.78
10    19.45
11    20.18
12    20.22
13    20.26
14    20.28
15    20.46
16    20.43
17    20.02
18    19.50
19    20.18
Name: night bedtime, dtype: float64

In [80]:
no_nap_bedtime = df['night bedtime'][df['napping'] == 0]
no_nap_bedtime

0    20.45
1    19.23
2    19.60
3    19.46
4    19.21
Name: night bedtime, dtype: float64

In [81]:
#nap bedtime mean
nap_bedtime_mean = nap_bedtime.mean()
nap_bedtime_mean

20.304

In [82]:
# no nap bedtime mean
no_nap_bedtime_mean = no_nap_bedtime.mean()
no_nap_bedtime_mean

19.590000000000003

**Question**: What is the sample difference of mean bedtime for nappers minus no nappers?

In [83]:
mean_dif = nap_bedtime_mean - no_nap_bedtime_mean
mean_dif

0.7139999999999951

Standard deviation for $X_{nap}$ and $X_{no\ nap}$.

In [84]:
# use ddof parameter 1 to get sample standard deviation
nap_s_bedtime = np.std(nap_bedtime, ddof = 1)
nap_s_bedtime

0.5910619981984009

In [85]:
no_nap_s_bedtime = np.std(no_nap_bedtime, ddof = 1)
no_nap_s_bedtime

0.5075923561284187

**Question**: What is the s.e.$(\bar{X}_{nap} - \bar{X}_{no\ nap})$?

We expect the variance in sleep time for toddlers who nap and toddlers who don't nap to be the same. So we use a pooled standard error.

Calculate the pooled standard error of $\bar{X}_{nap} - \bar{X}_{no\ nap}$ using the formula below.

$s.e.(\bar{X}_{nap} - \bar{X}_{no\ nap}) = \sqrt{\frac{(n_1-1)s_1^2 + (n_2-1)s_2^2}{n_1+n_2-2}(\frac{1}{n_1}+\frac{1}{n_2})}$

In [86]:
a = (15-1)*nap_s_bedtime**2
b = (5-1)*no_nap_s_bedtime**2
c = (1/15)+(1/5)
pooled_se = np.sqrt(((a+b)/18)*c)
pooled_se

0.2961871280370147

**Question**: Given our sample size of $n$, how many degrees of freedom ($df$) are there for the associated $t$ distribution?

Answer : 18


Now calculate the $t$-test statistic for our first hypothesis test using  
* pooled s.e.($\bar{X}_{nap} - \bar{X}_{no\ nap}$)  
* $\bar{X}_{nap} - \bar{X}_{no\ nap}$  
* $\mu_{0,\ nap} - \mu_{0,\ no\ nap}=0$, the population difference in means under the null hypothesis

In [87]:
tstats = (mean_dif - 0)/pooled_se
tstats

2.4106381824626966

**Question**: What is the p-value for the first hypothesis test?

For a discussion of probability density functions (PDF) and cumulative distribution functions (CDF) see:

https://integratedmlai.com/normal-distribution-an-introductory-guide-to-pdf-and-cdf/

To find the p-value, we can use the CDF for the t-distribution:
```
t.cdf(tstat, df)
```
Which for $X \sim t(df)$ returns $P(X \leq tstat)$.

Because of the symmetry of the $t$ distribution, we have that 
```
1 - t.cdf(tstat, df)
```
returns $P(X > tstat)$

The function `t.cdf(tstat, df)` will give you the same value as finding the one-tailed probability of `tstat` on a t-table with the specified degrees of freedom.

Use the function `t.cdf(tstat, df)` to find the p-value for the first hypothesis test.

In [88]:
1-t.cdf(tstats, 18) #t-statistic formula for two sided test

0.013417041438843036

**Question**: What are the t-statistic and p-value for the second hypothesis test?

Calculate the $t$ test statistics and corresponding p-value using the `scipy` function `scipy.stats.ttest_ind(a, b, equal_var=True)` and check with your answer. 

**Question**: Does `scipy.stats.ttest_ind` return values for a one-sided or one-sided test?

two-sided test

**Question**: Can you think of a way to recover the results you got using `1-t.cdf` from the p-value given by `scipy.stats.ttest_ind`?

Use the `scipy` function `scipy.stats.ttest_ind(a, b, equal_var=True)` to find the $t$ test statistic and corresponding p-value for the second hypothesis test.

In [89]:
from scipy import stats
import scipy

In [90]:
scipy.stats.ttest_ind(nap_bedtime, no_nap_bedtime, equal_var = True)

Ttest_indResult(statistic=2.4106381824626966, pvalue=0.026834082877686044)

In [91]:
pvalue = 0.026834082877686044 / 2
pvalue

0.013417041438843022

**Question**: For the $\alpha=.05$, do you reject or fail to reject the first hypothesis?

For the first hypothesis, p_value is 0.026 which is < alpha of 0.05. So we reject the null hypothesis. Therefore, night bedtime in children who nap is less or equal to the children who doesn't nap.

2. The average 24 h sleep duration (in minutes) for napping toddlers is different from toddlers who don't nap.

$$H_0: \mu_{nap}=\mu_{no\ nap}, \ H_a:\mu_{nap}\neq\mu_{no\ nap}$$
Or equivalently:
$$H_0: \mu_{nap}-\mu_{no\ nap}=0, \ H_a:\mu_{nap}-\mu_{no\ nap} \neq 0$$

First isolate `night bedtime` into two variables - one for toddlers who nap and one for toddlers who do not nap.

In [92]:
# Test for the 2nd hypothesis
nap_night_duration = df['night sleep duration'][df['napping'] == 1]
no_nap_night_duration = df['night sleep duration'][df['napping'] == 0]

In [93]:
nap_night_duration

5     602.20
6     618.40
7     526.80
8     626.80
9     549.50
10    655.20
11    611.20
12    611.80
13    618.80
14    548.00
15    593.25
16    649.80
17    573.60
18    693.40
19    615.33
Name: night sleep duration, dtype: float64

In [94]:

sleep_duration = df.groupby(['napping']).agg({'night sleep duration': [np.mean, np.std]})
sleep_duration

night sleep duration           
                        mean        std
napping                                
0.0                  675.040  56.169636
1.0                  606.272  43.858825

In [95]:

scipy.stats.ttest_ind(nap_night_duration, no_nap_night_duration, equal_var = True)

Ttest_indResult(statistic=-2.8409428070482057, pvalue=0.01084088459007456)

In [96]:
x = (15-1)*56.169636**2
y = (5-1)*43.858825**2
z = (1/15)+(1/5)
pooled_set = np.sqrt(((x+y)/18)*z)
pooled_set

27.719435149698352

In [97]:
ts = (606.272-675.040)/pooled_set

In [98]:
p_val=1-t.cdf(ts, 18)
p_val

0.9883937818260117

Question: For the  α=.05 , do you reject or fail to reject the second hypothesis?

H : The average 24 h sleep duration (in minutes) for napping toddlers is different from toddlers who don't nap.

For the second hypothesis, pvalue is 0.988 which is bigger than alpha of 0.05 so we fail to reject the null hypothesis and say that in children who nap and do not nap, there's difference of night sleep duration.


**More reading**
[How to use stats :](https:// [link text](https://)) https://www.programcreek.com/python/example/113287/scipy.stats.t.cdf

https://towardsdatascience.com/one-tailed-or-two-tailed-test-that-is-the-question-1283387f631c
